### PR TITLE
xcatprobe osdeploy add monitoring computes.log

### PR DESCRIPTION
**Pull request by xuweibj**
1. add monitoring computes.log
2. add dhcp message catch
3. if command without -n <noderange>, show error
4. modified method of getting macs of node

### CODE REVIEW

File: xCAT-probe/subcmds/osdeploy [hunk 1]
```diff
@@ -85,24 +81,44 @@ sub check_noderange{
...
         my $nic    = $4;
         my $record = "Receive DHCPREQUEST from $mac for $ip via $nic";
 
         if (exists $macmap{$mac}) {
-            my $node   = $macmap{$mac}{"node"};
+            my $node = $macmap{$mac}{"node"};
             probe_utils->send_msg("$output", "d", "$record");           
             push(@{ $rawdata{$node}{"history"} }, $record);
 
             if ($macmap{$mac}{"ip"} != $ip) {
```
Comment (daniceexi) [inline xCAT-probe/subcmds/osdeploy]: it's possible multiple nics can get IP during the netbooting. the correct way is to use the first mac which does not have NOIP

File: xCAT-probe/subcmds/osdeploy [hunk 2]
```diff
@@ -179,6 +221,31 @@ sub handle_cluster_msg {
     return 0;
 }
 
+sub handle_compute_msg {
+    my $line = shift;
+    my $sender;
+    my $node;
+    my $msg;
+
+    if ($line =~ /.+\d+:\d+:\d+\s+(.+)\s+(xcat.+)/i) {
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Will "\s*\d{2}:\d{2}:\d{2}\s* be more precise?

File: xCAT-probe/subcmds/osdeploy [hunk 3]
```diff
@@ -60,19 +62,20 @@ sub check_noderange{
...
+                if ($mac =~ /\!\*NOIP\*/) {
+                    $mac =~ s/\!\*NOIP\*//g;
+                    $macmap{$mac}{"ip"} = "NOIP";
+                }
+                else {
+                    $macmap{$mac}{"ip"} = $ip;
+                    $macmap{$mac}{"node"} = $currentnode;
+                }
+            }
+            $ip = "NOIP";
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: I do think you shall re-initial the $ip variable after line 62 when a new node found.

File: xCAT-probe/subcmds/osdeploy [hunk 4]
```diff
@@ -60,19 +62,20 @@ sub check_noderange{
...
-            $macmap{$2}{"node"} = $currentnode;
         } elsif ($_ =~ /^mac=(.+)/i) {
-            $macmap{$1}{"ip"} = $ip;
-            $macmap{$1}{"node"} = $currentnode;
+            $mac_line = $1;
+            @macs = split(/\|/, $mac_line);
+            foreach my $mac(@macs) {
+                if ($mac =~ /\!\*NOIP\*/) {
+                    $mac =~ s/\!\*NOIP\*//g;
+                    $macmap{$mac}{"ip"} = "NOIP";
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Do you also need to set "node" value for $macmap{$mac}?

File: xCAT-probe/subcmds/osdeploy [hunk 5]
```diff
@@ -85,24 +88,44 @@ sub check_noderange{
...
             push(@{ $rawdata{$node}{"history"} }, $record);
 
-            if ($macmap{$mac}{"ip"} != $ip) {
+            if ($macmap{$mac}{"ip"} != "NOIP" and $macmap{$mac}{"ip"} != $ip) {
                 my $warn_msg = "The ip of $mac from DHCP $ip is different with definition $macmap{$mac}{'ip'}.";
                 probe_utils->send_msg("$output", "w", "$warn_msg");
-                if (exists($rawdata{$node})) {
-                    push(@{ $rawdata{$node}{"history"} }, $warn_msg);
-                }
+                push(@{ $rawdata{$node}{"history"} }, $warn_msg);
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Won't you shall update the value of "$macmap{$mac}{ip}"?

File: xCAT-probe/subcmds/osdeploy [hunk 6]
```diff
@@ -113,13 +136,43 @@ sub handle_dhcp_msg {
...
             }
         }
+    } elsif ($msg =~ /.+BOOTREQUEST\s+from\s+(.+)\s+via\s+([^:]+)(.*)/) {
+        my $mac = $1;
+        my $nic = $2;
+        my $record = "Receive BOOTREQUEST from $mac via $nic";
+
+        if (exists $macmap{$mac}) {
+            probe_utils->send_msg("$output", "d", "$record");
+            push(@{ $rawdata{$node}{"history"} }, $record);
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: How does $node get value?

File: xCAT-probe/subcmds/osdeploy [hunk 7]
```diff
@@ -166,16 +219,44 @@ sub handle_cluster_msg {
         $status = $2;
 
         if (exists($rawdata{$node})) {
-            my $record = "Recv from $node : xcat: status is $status";
+            my $record = "Receive from $node : status is $status";
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Also for line 211?

**Comment by zet809** [comment]
Looks good to me, I agree to merge.

**Pull request merged**